model/src/util/image_hash.rs

```diff
@@ -0,0 +1,775 @@
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let input = "1acefe340fafb4ecefae407f3abdb323";
+    /// let parsed = ImageHash::parse(input.as_bytes())?;
+    ///
+    /// let (bytes, is_animated) = (parsed.bytes(), parsed.is_animated());
+    ///
+    /// let constructed = ImageHash::new(bytes, is_animated);
+    /// assert_eq!(input, constructed.to_string());
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// [`is_animated`]: Self::is_animated
+    /// [`bytes`]: Self::bytes
+    pub const fn new(bytes: [u8; 16], animated: bool) -> Self {
+        Self { animated, bytes }
+    }
+
+    /// Parse an image hash into an efficient integer-based storage.
+    ///
+    /// # Examples
+    ///
+    /// Parse a static image hash:
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let input = "b2a6536641da91a0b59bd66557c56c36";
+    /// let parsed = ImageHash::parse(input.as_bytes())?;
+    ///
+    /// assert!(!parsed.is_animated());
+    /// assert_eq!(input, parsed.to_string());
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// Parse an animated image hash:
+    ///
+    /// ```
+    /// # fn main() -> Result<(), Box<dyn std::error::Error>> {
+    /// use twilight_model::util::ImageHash;
+    ///
+    /// let input = "a_b2a6536641da91a0b59bd66557c56c36";
+    /// let parsed = ImageHash::parse(input.as_bytes())?;
+    ///
+    /// assert!(parsed.is_animated());
+    /// assert_eq!(input, parsed.to_string());
+    /// # Ok(()) }
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Returns an [`ImageHashParseErrorType::Format`] error type if the
+    /// provided value isn't in a Discord image hash format. Refer to the
+    /// variant's documentation for more details.
+    ///
+    /// Returns an [`ImageHashParseErrorType::Range`] error type if one of the
+    /// hex values is outside of the accepted range. Refer to the variant's
+    /// documentation for more details.
+    pub const fn parse(value: &[u8]) -> Result<Self, ImageHashParseError> {
+        let animated = Self::starts_with(value, ANIMATED_KEY.as_bytes());
+
+        let mut seeking_idx = if animated { ANIMATED_KEY.len() } else { 0 };
+        let mut storage_idx = 0;
+
+        if value.len() - seeking_idx != HASH_LEN {
+            return Err(ImageHashParseError::format());
+        }
+
+        let mut bits = 0;
+
+        while seeking_idx < value.len() {
+            let byte = match value[seeking_idx] {
+                byte @ b'0'..=b'9' => byte - b'0',
+                byte @ b'a'..=b'f' => byte - b'0' - 39,
+                other => return Err(ImageHashParseError::range(seeking_idx, other)),
+            };
+
+            bits |= (byte as u128) << 124_usize.saturating_sub(storage_idx * 4);
+            seeking_idx += 1;
+            storage_idx += 1;
+        }
+
+        Ok(Self {
+            animated,
+            bytes: bits.to_be_bytes(),
```

Comment:
Is there a reason this is BE instead of NE?

Reply:
not really

Reply:
I guess its just a nit then since this is just going to be overhead on most systems Twilight will be used on.

Reply:
Then probably do a `#[cfg]` check on the target endian? `.to_be_bytes` for Big-Endian platforms, and `.to_le_bytes` for Little-Endian platforms - but I believe `.to_ne_bytes` already handles this.